### DISCUSSION TITLE
Dynamically switching template while editing - concerns and thoughts

### DISCUSSION THREAD

**Discussion by mattsherman** [woocommerce/woocommerce#40399] [WooCommerce New Product Editor]
I'm concerned about the potential complexities and issues that may arise if we support dynamically switching the template (under the hood, without an explicit user decision to do this) while editing a product. I've shared some of these concerns internally on team calls, but I thought it would be valuable to put them down in a discussion so that we can more easily discuss them!

## Background

The block template API, and the related template registry support multiple templates for the product editor.

We are currently hardcoding the implementation to use a single template, `SimpleProductTemplate`.

However, the current plan is to extend things to allow for a different template to be assigned to each product type (simple, variable, grouped, etc.)

We are also planning to allow the template to be dynamically switched while in the editor based on the state of the edited product. This would happen without the user explicitly deciding to switch the template (or the product type). This ability to switch templates "under-the-hood" would be available to both core and extensions.

One specific near-term scenario where we are planning on making use of this is to switch from the "simple product" template to the "variable product" template when the user adds variation options to the edited product (without having the user explicitly switch the product type to "variable").

## Concerns

If we support dynamically switching the template (under-the-hood) while editing a product...

- There is no guarantee that the new template bares any resemblance to the current template. The order of groups, sections, and fields could be completely different.
    - This could be quite jarring to the user.
    - This makes smoothly animating the showing/hiding of groups/sections/fields quite difficult (we could do some sort of a diff between the templates and determine which fields are being shown/hidden).
- Even if the templates are largely similar, there could be subtle, inadvertent differences (slightly different labels, etc.) due to a number of reasons:
    - Forgetting to update the implementation of both the templates when changes are made. There are techniques that can be used to share implementation between templates, but there is no guarantee that this has been done.
    - Extensions mistakenly only targeting a single template instead of both the templates. Again, there are techniques we provide that make it easy for an extension to target all product editor templates, but nothing guarantees this is done.
- There is no guarantee that the "same" fields in both template have the same template block id (say the product name field), parent container, order, etc.
    - This makes it more difficult for extensions to make template modifications that are based on other blocks (support recently introduced in #40139). While extensions should always be coded robustly to handle scenarios where blocks they are dependent on are missing or not where they thought they would be, it is even more important when there can be multiple very similar templates.

## My thoughts

- Switching the template dynamically while editing a product seems to only be a reasonable option if the two templates are very similar, with only a few fields shown/hidden that differ (this is the case with the "simple" vs "variable" product types).
    - However, if that is the case, I believe that it would be both easier from an implementation standpoint, and better from a UX standpoint to make use of the upcoming conditional visibility support (see #39434) due to the concerns shared above.
- We should strongly discourage, or not even support, dynamically switching the template while editing a product.
    - If we support this, the guideline should be that this only happen in response to an explicit decision by the user to switch the template or change the product type, so that the user is aware of the potential for drastic change.

## Your thoughts?

Please share your thoughts on this issue by the end of the week, Friday, September 29 so that we can determine the best way to proceed!

/cc @woocommerce/mothra

**Comment by mattsherman** [comment]
Another concern I just thought of...

Let's say extension A decides to dynamically switch to template Foo when condition X is true.

Now, another extension B decides to dynamically switch to template Bar when condition X is true.

Which template is shown?

If template Foo and template Bar are largely similar to the base template, they may be able to co-exist easier if they just conditionally hide fields on the base template that weren't needed when condition X is true and show their custom fields. But, if they are both switching to an entirely different template, this wouldn't work.

**Comment by nathanss** [comment]
I would love it if we could "get away" without dynamically switching the template while editing the product. In the "simple vs variable" scenario, it just doesn't feel like the right abstraction to me, since they are very similar (of course, we would have to give a good name to the "Simple & Variable" template). Switching the template seems almost like a full-page refresh to me.

It seems like, between product types and traits, we're struggling to find the guidelines on where to use a template. In my mind, I would use different templates for scenarios that are very different from each other, and it would be best if the template was "selected" as a first step. We could allow developers to add more templates to this first step. I found this in the design:

<img width="555" alt="image" src="https://github.com/woocommerce/woocommerce/assets/13437655/2b6e4233-ef4e-472e-b5b6-afc19c5795d5">


 I agree with preferring conditionally showing/hiding fields whenever possible.

**Comment by louwie17** [comment]
Thanks for starting this discussion here @mattsherman!
I agree with your concerns and your solutions, that if only small changes happen to a template we should stick with the conditional rendering, and for large differences that we should have a clear visual transition so users know when they are switching templates.

Although I would say that we could still have some similar templates like variable vs simple, but make sure we still add a visual transition.
At the moment I was the one that add some logic in-behind the scenes that automatically changes a product to `variable` if variations are added through the UI, although I am not a huge fan of this.
It would be great if even then, we show a modal of some sort, saying: `You are now editing a variable product`. On top of this, it may be worth having a visual `tag` somewhere that shows what type of template is being rendered, by the title of the template (if that is available).

cc: @jarekmorawski would love your feedback on this discussion.

**Reply by mattsherman** [reply]
> On top of this, it may be worth having a visual tag somewhere that shows what type of template is being rendered, by the title of the template (if that is available).

Do you mean always visible to the end user, or as a debugging aid? 

If the templates can be selected by the end user, and especially if they are not 1:1 with product types, then I would agree that the template name should be visible somewhere. If not, then it could probably just be a debugging aid.

**Reply by louwie17** [reply]
I was thinking visible to the user. But your second line sums it up pretty well:

> If the templates can be selected by the end user, and especially if they are not 1:1 with product types, then I would agree that the template name should be visible somewhere. If not, then it could probably just be a debugging aid.

**Reply by AnnaMag** [reply]
> especially if they are not 1:1 with product types

I assume that "product types" refers to the default, custom (e.g. defined and saved by users), and traits (e.g. digital). Could you help me understand when the lack of 1:1 correspondence could occur?

**Comment by AnnaMag** [comment]
Thanks @mattsherman.

Could we cover all default products and traits with conditional rendering?

To keep things flexible to account for custom product types, I would +1 creating templates, but controlling their usage (non-dynamic, e.g. creating/editing from template).

**Reply by mattsherman** [reply]
> Could we cover all default products and traits with conditional rendering?

Traits, 100%... in fact, we need to handle them with conditional visibility since traits are not mutually exclusive.

Default product types... 100% that we could handle it with conditional visibility. For Grouped and External/Affiliate products, we do make a number of visibility changes. It can be handled, but I do wonder for those if separate templates could be better from a UX perspective, since they feel more distinctly different.
